### DISCUSSION TITLE
Proposal:  Implicit "select"

### DISCUSSION THREAD

**Discussion by bondsbw** [dotnet/csharplang#5170] [Language Ideas]
Relax the requirement to end every LINQ query expression with a `select` clause.  An implicit `select` would form a tuple containing all range variables in scope at the end of the LINQ expression.

### Simple example
New:
```c#
var lowStockProducts = 
    from product in Products
    let unitsAvailable = product.UnitsInStock - product.UnitsOnOrder
    where unitsAvailable < 5;
```
Equivalent to:
```c#
var lowStockProducts = 
    from product in Products
    let unitsAvailable = product.UnitsInStock - product.UnitsOnOrder
    where unitsAvailable < 5
    select (product, unitsAvailable);
```

### More complex example
```c#
var categorizedProducts =
    from product in ps
    group product by product.CategoryID into grouped
    let mostExpensiveProducts =
        from product2 in grouped
        where product2.UnitPrice == grouped.Max(product3 => product3.UnitPrice)
        select product2
    join category in cs on grouped.Key equals category.CategoryID;
```
`group ... by ... into` removes previous range variables from scope, so the implied `select` clause would not include `product`:
```c#
    select (grouped, mostExpensiveProducts, category)
```
Note that nested queries would still require `select` in order to resolve ambiguity.

**Comment by DavidArno** [comment]
> Relax the requirement to end every LINQ query expression with a select clause.

Yes, I agree.

> An implicit select would form a tuple containing all range variables in scope at the end of the LINQ expression.

No, that's horrible in my view. The natural assumption for me is that the equivalent of:
```cs
var lowStockProducts = 
    from product in Products
    let unitsAvailable = product.UnitsInStock - product.UnitsOnOrder
    where unitsAvailable < 5;
```
would be:
```cs
var lowStockProducts = 
    from product in Products
    where unitsAvailable < 5
    select product;
```
with a warning that `unitsAvailable` is unused. In other words I like the idea of dropping the `select` but only if, in this case, an `IEnumerable<Product>` is returned.

**Comment by CyrusNajmabadi** [comment]
Interesting.  I like this.  With our inferred tuple name work too, you'd get a nice ```IEnumerable<(Product product, int unitsAvailable)>``` result.

**Comment by CyrusNajmabadi** [comment]
Another way we could potentially have this fall out would be to make it so that transparent identifiers were not compiled into anonymous types, but instead used tuples.  We'd then just allow you to work with that tuple, instead of forcing projection into some final form.  

Actually, using tuples for transparent identifiers might just be desirable anyways from a perf perspective...  hrmm...

**Comment by HaloFour** [comment]
> Actually, using tuples for transparent identifiers might just be desirable anyways from a perf perspective...

That came up on the Roslyn repo.  The actual projection being an implementation detail changing the type used could be done without modifying the spec in any way and most people would likely not know the difference.

However, if omitting `select` did include the range values that would cause that intermediate form to leak out and thus require that behavior to be standardized in the specification.

**Comment by jnm2** [comment]
At least for now we could allow this if there is a single range variable. That would encourage me to use LINQ syntax more often. Way more often actually. That may be the number one reason I switch to method syntax.

(Edit: that, and not liking to backtrack to `).ToList()` or `).ToDictionary(...)`.)

**Comment by alrz** [comment]
I think this should not catch all range variables in a tuple, I'd prefer what VB does,

```cs
var q = 
  from item in list
  let foo = item.foo
  where foo > 0;

->

var q = list
  .Select(item => (item, foo: item.foo))
  .Where(t => t.foo > 0);
```
No magic involved.

**Comment by bondsbw** [comment]
@jnm2 
> At least for now we could allow this if there is a single range variable. 

The single range variable case is indeed important and very common.

_(To the group)_ In such a scenario would it return `IEnumerable<Product>` or `IEnumerable<(Product product,)>` (oneple)?  I like the former but would the disparity with the multi-value (tuple) case be a problem?  

@alrz 

I'm a bit confused by your example.  Doesn't that query capture all the range variables?  The result is `IEnumerable<ValueTuple<TItem, TFoo>>` (will be `IEnumerable<(TItem item, TFoo foo)>`).

**Comment by jnm2** [comment]
If there is a single range variable I absolutely need a result of `IEnumerable<Product>` or else I'll have to put the `select` right back in and I'll be back to square one.

**Comment by alrz** [comment]
@bondsbw

yeah, I was incorrect on VB behavior, I just don't like the level of implicitness, maybe it's ok for VB.

**Comment by bondsbw** [comment]
@CyrusNajmabadi 

> Interesting. I like this. With our inferred tuple name work too, you'd get a nice IEnumerable<(Product product, int unitsAvailable)> result.

When the result is used later, would that feature pull the inferred names directly into scope:

```c#
var q = 
  from item in list
  let foo = item.foo
  where foo > 0;

foreach (var _ in q)
{
    Console.WriteLine(item);
    Console.WriteLine(foo);
}
```
Or would the foreach block necessarily need to include a dummy identifier:
```c#
foreach (var x in q)
{
    Console.WriteLine(x.item);
    Console.WriteLine(x.foo);
}
```

**Comment by jnm2** [comment]
@bondsbw 

```c#
foreach (var (item, foo) in q)
{
    Console.WriteLine(item);
    Console.WriteLine(foo);
}
```

😁

**Comment by bondsbw** [comment]
@jnm2 But we would get that _without_ inferred tuple names.  Where's the fun in that?  :stuck_out_tongue_winking_eye:

**Comment by jnm2** [comment]
True. I really don't like the discard syntax or the magic though. :-(

**Comment by CyrusNajmabadi** [comment]
> However, if omitting select did include the range values that would cause that intermediate form to leak out and thus require that behavior to be standardized in the specification.

Omitting the select would not cause the intermediate form to leak out.  It would simply codify what you get when you don't have a select.  You could still have whatever intermediate form you wanted, and you'd just translate (only if necessary) to the spec'ed form.

Now, from an impl perspective, it would probably be desirable for the intermediate form to be hte same.  but there's be no requirement.  It would be as if you just ended every non-select-ending query with:

```select (all, your, range, variables)```.

**Comment by CyrusNajmabadi** [comment]
Sorry, i never meant to imply that a oneple would ever be created.  If there was one range variable, you'd just select that out.  If there were multiple, you'd get a tuple with those range variables.

**Comment by marafiq** [comment]
~~~
var lowStockProductsQuery = 
    from product in Products
    join cat in categories on product.categoryId=cat.Id 
    let unitsAvailable = product.UnitsInStock - product.UnitsOnOrder
    where unitsAvailable < 5
~~~

With implicit select, lowStockProductQuery can be IQueryable<(tuple of all in scope ranges or anonymous type with range props)> or IEnumerable<...>. Advantage is that you can reuse the query which has the core logic.

like 

~~~
lowStockProductsQuery .Select(x=> new {product.Name,cat.Name})
~~~
OR
~~~
lowStockProductsQuery .Select(x=> new {product.Name,cat.Name,cat.Id,Product.Id})
~~~

Currently query must end with `select` or `group by`, with that you can explicitly select manually i.e. `new {cat,product,unitsAvailable}` and the re-use it. But in EF that will result in sub-query which is not required & bad for performance always selecting all fields.

Its common to have queries written with joins but you have to duplicate it whole query to select different fields since select does not accept Func<T..>. This will remove all that duplication.

**Comment by CyrusNajmabadi** [comment]
> But in EF that will result in sub-query which is not required & bad for performance always selecting all fields.

That sounds like an issue for EF to deal with :)

**Reply by marafiq** [reply]
Since its part of expression tree with explicit syntax, EF can not do much there. I think?

Is it difficult to consider this alone, how to make the queries reusable.

**Reply by CyrusNajmabadi** [reply]
>EF can not do much there. I think?

EF can do whatever it wants here.  It can examine the Select and decide if it's not something it actually needs to do work for and which would equivalent to whatever semantics it wants.

**Reply by marafiq** [reply]
> EF can do whatever it wants here.  It can examine the Select and decide if it's not something it actually needs to do work for and which would equivalent to whatever semantics it wants.

That seems doable in sense of outer select does not have any conditions attached to it, simply can be stripped out. Will ask there. 

It seems like you don't see value in re-usability part of query syntax by relaxing select or group condition or allowing select part to accept delegates?

**Reply by CyrusNajmabadi** [reply]
> It seems like you don't see value in re-usability part of query syntax by relaxing select or group condition or allowing select part to accept delegates?

As per my messages above in the entire thread, i think the idea is generally ok.  But i also think there's very little need for it, and it would not meet my bar in terms of getting attention vs the other things currently on our plate.  However, that could always change in the future! :)

**Reply by marafiq** [reply]
I appreciate you taking time respond. 

As an application dev, I feel differently as it improves things by lot. Especially when you dealing to refactor lot sql stored procs to query syntax.  

Though is there metric or voting system used to prioritize issues given as u said if idea is generally okay. Or it's up to core team to pick stuff.

**Reply by CyrusNajmabadi** [reply]
> I appreciate you taking time respond.

Definitely!

> As an application dev, I feel differently as it improves things by lot. Especially when you dealing to refactor lot sql stored procs to query syntax.

Understood :)

> Though is there metric or voting system used to prioritize issues given as u said if idea is generally okay. 

There are many factors that feed into our decisions.  Voting can sway things.  However, currently this doesn't have a lot of interest:

![image](https://user-images.githubusercontent.com/4564579/132765746-962b35b8-e65e-40f3-a58a-dc777bf2457f.png)

> Or it's up to core team to pick stuff.

The language team always has the final say and final word in all of this.  However, community feedback does feed heavily into our decision making.
